crates/rpc/src/jsonrpc/request.rs

```diff
@@ -0,0 +1,241 @@
+use serde::Deserialize;
+use serde_json::value::RawValue;
+
+use crate::jsonrpc::{RequestId, RpcError};
+
+use std::borrow::Cow;
+
+#[derive(Debug)]
+pub struct RpcRequest<'a> {
+    pub method: Cow<'a, str>,
+    pub params: RawParams<'a>,
+    pub id: RequestId<'a>,
+}
+
+#[derive(Debug, Default, Deserialize)]
+pub struct RawParams<'a>(#[serde(borrow)] Option<&'a RawValue>);
+
+impl<'a> RawParams<'a> {
+    /// Returns true if there are no params or the list of params is empty.
+    pub fn is_empty(&self) -> bool {
+        let Some(params) = self.0 else {
+            return true;
+        };
+
+        let params = params.get().trim().as_bytes();
```

Comment:
Since json is by default utf-8 it'd be safer to do these checks on str slices rather than just bytes.

Then you could also check for all utf-8 whitespace chars instead of just ascii (https://doc.rust-lang.org/std/primitive.char.html#method.is_whitespace).

Reply:
I think I actually did this purposefully because I thought speed was more important and jsonrpsee isn't doing char checks as well. But sure we can do utf-8 instead although I'd find it weird to have non-ascii whitespace spread around..

Reply:
Yeah, speed matters more.